server/py/framework/utils/clients/discovery.py

```diff
@@ -0,0 +1,150 @@
+        self, method: str, path: str
+    ) -> typing.Optional[ServiceInstance]:
+        """
+        Resolve path and returns the matching service instance for the request.
+
+        :param method: HTTP method of the request
+        :param path: URL path to match against service patterns
+        :return: ServiceInstance matching the request or None if no match
+        """
+        method = method.lower()
+        if service_name := self._find_service(method, path):
+            return self.get_service(service_name)
+        return None
+
+    def get_service(self, service_name: str) -> typing.Optional[ServiceInstance]:
+        """Get all healthy instances of a service."""
+        return self.services.get(service_name, None)
+
+    def _resolve_service_method_routes(
+        self, service_name: str
+    ) -> dict[str, list[re.Pattern]]:
+        """Resolve service routes per method for a service"""
+        method_routes = {
+            "get": [],
+            "post": [],
+            "put": [],
+            "delete": [],
+        }
+        routes = self._service_routes(service_name)
+        for methods, path in routes:
+            if methods == ["*"]:
+                for method in method_routes:
+                    method_routes[method].append(re.compile(path))
+                continue
+
+            for method in methods:
+                method_routes[method].append(re.compile(path))
+
+        return method_routes
+
+    def _find_service(self, method: str, path: str):
+        """
+        Find first service matching the given request URL
+
+        :param method: HTTP method of the request
+        :param path: URL path to match against service patterns
+        :return: Name of matching service or None if no match
+        """
+        for service_name, service_instance in self.services.items():
+            routes_patterns = service_instance.method_routes[method]
+            for route_pattern in routes_patterns:
+                if route_pattern.match(path):
+                    return service_name
+        return None
+
+    @staticmethod
+    def _resolve_service_url(service_name: str) -> str:
+        """Resolve service URL by service name."""
+        return f"http://mlrun-{service_name}.{mlconf.namespace}.svc.cluster.local:8080"
+
+    @staticmethod
+    def _service_routes(service_name: str) -> list:
+        """Get all routes for a service."""
+        return {
+            "api-chief": [
+                # Alerts routes that require chief capabilities
+                (["put", "delete"], "alert_templates.*"),
+                (["*"], "projects/.+/alerts.*"),
+                (["post"], "projects/.+/events.*"),
+            ],
+            "api": [],
+            "alerts": [
+                (["put", "get", "delete"], "alert_templates.*"),
+                (["*"], "projects/.+/alerts.*"),
+                (["post"], "projects/.+/events.*"),
+            ],
+        }[service_name]
```

Comment:
Again I think it is a bit strange that the discovery service "know" the microservices under it.

Reply:
Same about dummy discovery. I agree this is not the best place for defining the routed. I don't have a better idea atm. This should probably be held by the API service somewhere